### ISSUE TITLE
Problem: Lot of "Archivematica Fail Report for Transfer" emails sent when upgrading from AM1.10 to AM1.11

### ISSUE THREAD

**Issue by mamedin** [archivematica/Issues#1238]
**Expected behaviour**

No notification email should be sent if there was no pending transfer or red bubble in the transfer or ingest tags

**Current behaviour**

We have upgraded 4 customers' AM servers from AM1.10 to AM1.11 and this issue has happened in 3 of these 4 upgrades. One of them has sent more than 500 emails.

**Steps to reproduce**

Upgrade AM from AM1.10 to AM1.11 in a production server

**Your environment (version of Archivematica, operating system, other relevant details)**

It happened in several environments:

* Centos 7 deployed from packages when upgrading from AM1.10 to AM1.11.1
* Centos 7 deployed from sources when upgrading from v1.10.1 --> dev/issue-1211-cas-authentication --> qa/1.12.x-sfu-dashboard-features
* Ubuntu Xenial from sources when upgrading from AM1.10 to v1.11.2

*For Artefactual use:**

Before you close this issue, you must check off the following:

- [ ] All pull requests related to this issue are properly linked
- [ ] All pull requests related to this issue have been merged
- [ ] A testing plan for this issue has been implemented and passed (testing plan information should be included in the issue body or comments)
- [ ] Documentation regarding this issue has been written and merged (if applicable)
- [ ] Details about this issue have been added to the release notes (if applicable)

**Comment by mamedin** [comment]
This is the mail sent:

**Subject**: Archivematica Fail Report for Transfer: XXXXXXXXXX-4bccd226-3826-45e2-960a-5676d611c03f



Unit type | Total time processing | Total file size | Number of files | Average file size KB | Average file size MB
-- | -- | -- | -- | -- | --
Transfer | 00:29:28 | 39 | 87679832 | 2248.20082051 | 2.24820082

Type | Status | Started
-- | -- | --
Check transfer directory for objects | 4 | 2020-06-19 09:35:58.406673+00:00
Check transfer directory for objects | 2 | 2020-06-19 08:58:17.333638+00:00
Check transfer directory for objects | 2 | 2019-11-06 15:02:16.530655+00:00
Move to SIP creation directory for completed transfers | 2 | 2019-11-06 15:02:15.623181+00:00
Create SIP from transfer objects | 2 | 2019-11-06 15:02:14.816127+00:00
Serialize Dublin Core metadata to disk | 2 | 2019-11-06 15:02:14.412710+00:00
Move to processing directory | 2 | 2019-11-06 15:02:13.887989+00:00
Create SIP(s) | 2 | 2019-11-06 15:02:13.850768+00:00
Load options to create SIPs | 2 | 2019-11-06 15:02:13.808740+00:00
Check transfer directory for objects | 2 | 2019-11-06 15:02:13.520233+00:00
Move to SIP creation directory for completed transfers | 2 | 2019-11-06 15:02:13.201888+00:00
Create transfer metadata XML | 2 | 2019-11-06 15:02:12.801009+00:00
Parse external METS | 2 | 2019-11-06 15:02:12.464106+00:00
Check for specialized processing | 2 | 2019-11-06 15:02:12.426289+00:00
Examine contents? | 2 | 2019-11-06 15:02:12.107712+00:00
Move to examine contents | 2 | 2019-11-06 15:02:11.940237+00:00
Perform policy checks on originals? | 2 | 2019-11-06 15:02:11.832278+00:00
Determine if Dataverse METS XML needs to be parsed | 2 | 2019-11-06 15:02:11.788645+00:00
Generate METS.xml document | 2 | 2019-11-06 15:02:10.585460+00:00
Validate formats | 2 | 2019-11-06 15:01:42.651868+00:00
Load labels from metadata/file_labels.csv | 2 | 2019-11-06 15:01:42.045653+00:00
Characterize and extract metadata | 2 | 2019-11-06 15:01:36.016335+00:00
Load rights | 2 | 2019-11-06 15:01:35.615205+00:00
Store file modification dates | 2 | 2019-11-06 15:01:35.208936+00:00
Determine if transfer contains packages | 2 | 2019-11-06 15:01:34.942846+00:00
Move to extract packages | 2 | 2019-11-06 15:01:33.801919+00:00
Identify file format | 2 | 2019-11-06 15:01:29.336838+00:00
Determine which files to identify | 2 | 2019-11-06 15:01:29.299147+00:00
Do you want to perform file format identification? | 2 | 2019-11-06 15:01:29.135216+00:00
Move to select file ID tool | 2 | 2019-11-06 15:01:28.137333+00:00
Sanitize Transfer name | 2 | 2019-11-06 15:01:27.170808+00:00
Sanitize object's file and directory names | 2 | 2019-11-06 15:01:26.406761+00:00
Generate transfer structure report | 2 | 2019-11-06 15:01:26.364914+00:00
Move to processing directory | 2 | 2019-11-06 15:01:25.652865+00:00
Move to generate transfer tree | 2 | 2019-11-06 15:01:24.757455+00:00
Scan for viruses | 2 | 2019-11-06 15:01:20.523060+00:00
Move to processing directory | 2 | 2019-11-06 15:01:20.257220+00:00
Workflow decision - send transfer to quarantine | 2 | 2019-11-06 15:01:20.209206+00:00
Find type to process as | 2 | 2019-11-06 15:01:20.168549+00:00
Move to workFlowDecisions-quarantineSIP directory | 2 | 2019-11-06 15:01:18.155620+00:00
Designate to process as a standard transfer | 2 | 2019-11-06 15:01:18.110786+00:00
Verify metadata directory checksums | 2 | 2019-11-06 15:01:17.684966+00:00
Process transfer JSON metadata | 2 | 2019-11-06 15:01:17.101665+00:00
Assign checksums and file sizes to objects | 2 | 2019-11-06 15:01:14.070594+00:00
Assign file UUIDs to objects | 2 | 2019-11-06 15:01:06.794211+00:00
Assign UUIDs to directories | 2 | 2019-11-06 15:01:06.347846+00:00
Assign UUIDs to directories? | 2 | 2019-11-06 15:01:06.321333+00:00
Include default Transfer processingMCP.xml | 2 | 2019-11-06 15:01:05.230127+00:00
Rename with transfer UUID | 2 | 2019-11-06 15:01:04.021257+00:00
Verify mets_structmap.xml compliance | 2 | 2019-11-06 15:01:03.223390+00:00
Verify transfer compliance | 2 | 2019-11-06 15:00:55.302459+00:00
Attempt restructure for compliance | 2 | 2019-11-06 15:00:54.410742+00:00
Determine if Dataverse conversion needs to happen | 2 | 2019-11-06 15:00:54.372203+00:00
Remove unneeded files | 2 | 2019-11-06 15:00:53.371857+00:00
Remove hidden files and directories | 2 | 2019-11-06 15:00:52.758688+00:00
Set transfer type: Standard | 2 | 2019-11-06 15:00:52.409115+00:00

**Comment by mamedin** [comment]
Looking at MCPServer logs, last lines for the transfer of the above example (AM1.10, last year):

``` 
MCPServer/MCPServer.log.2:INFO      2019-11-06 15:02:16  archivematica.mcp.server:unitTransfer:__init__:50:  Using existing Transfer 4bccd226-3826-45e2-960a-5676d611c03f at %sharedPath%watchedDirectories/SIPCreation/completedTransfers/XXXXXXXXXXXX-4bccd226-3826-45e2-960a-5676d611c03f/
MCPServer/MCPServer.log.2:INFO      2019-11-06 15:02:16  archivematica.mcp.server:jobChainLink:__init__:52:  Running Check transfer directory for objects (unit 4bccd226-3826-45e2-960a-5676d611c03f)
```

Logs for this transfer on AM1.11 upgrade:

``` 
MCPServer/MCPServer.log:INFO      2020-06-19 08:58:16  archivematica.mcp.server:unitTransfer:__init__:50:  Using existing Transfer 4bccd226-3826-45e2-960a-5676d611c03f at %sharedPath%watchedDirectories/SIPCreation/completedTransfers/XXXXXXXXXX-4bccd226-3826-45e2-960a-5676d611c03f/
MCPServer/MCPServer.log:INFO      2020-06-19 08:58:17  archivematica.mcp.server:jobChainLink:__init__:52:  Running Check transfer directory for objects (unit 4bccd226-3826-45e2-960a-5676d611c03f)
MCPServer/MCPServer.log:INFO      2020-06-19 09:35:58  archivematica.mcp.server.packages:packages:get_or_create_from_db_by_path:808:  Transfer 4bccd226-3826-45e2-960a-5676d611c03f updated (%sharedPath%watchedDirectories/SIPCreation/completedTransfers/RMA-2019-17_2014_Football_Meet_Greet-4bccd226-3826-45e2-960a-5676d611c03f/)
MCPServer/MCPServer.log:INFO      2020-06-19 09:37:23  archivematica.mcp.server.jobs.client:client:run:113:  Running Check transfer directory for objects (package 4bccd226-3826-45e2-960a-5676d611c03f)
MCPServer/MCPServer.log:INFO      2020-06-19 09:37:24  archivematica.mcp.server.jobs.client:client:run:113:  Running Email fail report (package 4bccd226-3826-45e2-960a-5676d611c03f)
MCPServer/MCPServer.log:INFO      2020-06-19 09:37:38  archivematica.mcp.server.jobs.client:client:run:113:  Running Cleanup failed Transfer (package 4bccd226-3826-45e2-960a-5676d611c03f)
MCPServer/MCPServer.log:INFO      2020-06-19 09:37:38  archivematica.mcp.server.jobs.client:client:run:113:  Running Move to the failed directory (package 4bccd226-3826-45e2-960a-5676d611c03f)
```

**Comment by replaceafill** [comment]
I've been able to reproduce this behaviour in the docker-compose environment:

![Screenshot_2020-06-19 Archivematica Dashboard - Transfer](https://user-images.githubusercontent.com/560781/85179156-77bf5b00-b23d-11ea-8112-a385e2a55378.png)

I also double checked the behaviour I found in the logs with Miguel's test VM.

The sequence of the problem seems to be:

1. In `1.10` completed transfers leave files in the `watchedDirectories/SIPCreation/completedTransfers` as reported in https://github.com/archivematica/Issues/issues/769:

    ```console
    home/replaceafill/.am/am-pipeline-data/watchedDirectories/SIPCreation/completedTransfers/
    └── demo_transfer_absolute-a11afdbd-17b2-45f2-8d83-4857b592505b
        ├── logs
        │   ├── bulk-4e1c6019-36d1-475c-a3a1-e4b083c7bda0
        │   │   ├── domain_histogram.txt
        │   │   ├── domain.txt
        │   │   └── report.xml
        │   ├── bulk-6ca26140-a934-4bfd-95fe-6b4e0c1d8104
        │   │   └── report.xml
        │   ├── bulk-74ae138b-2e39-4827-b5ae-2d20ee37aa69
        │   │   ├── exif.txt
        │   │   └── report.xml
        │   ├── bulk-77b61196-e045-4ef5-b58d-1e42714703fd
        │   │   ├── ccn_histogram.txt
        │   │   ├── ccn.txt
        │   │   └── report.xml
        │   ├── bulk-9f66814c-be8c-4535-bf41-a4e8aa3bbd61
        │   │   └── report.xml
        │   ├── bulk-c075978c-1814-4bf4-8a63-ab73680a1616
        │   │   └── report.xml
        │   ├── bulk-ee2c6608-6e28-4a23-b219-94b76b74b2f2
        │   │   ├── domain_histogram.txt
        │   │   ├── domain.txt
        │   │   ├── email_domain_histogram.txt
        │   │   ├── email_histogram.txt
        │   │   ├── email.txt
        │   │   ├── exif.txt
        │   │   └── report.xml
        │   ├── fileFormatIdentification.log
        │   ├── fileMeta
        │   └── filenameCleanup.log
        ├── metadata
        │   ├── checksum.md5
        │   ├── checksum.sha1
        │   ├── checksum.sha256
        │   ├── checksum.sha512
        │   ├── directory_tree.txt
        │   ├── identifiers.json
        │   ├── metadata.csv
        │   ├── rights.csv
        │   └── submissionDocumentation
        │       ├── Agreement-Gift-MBRS-project.pdf
        │       └── METS.xml
        ├── objects
        └── processingMCP.xml

    13 directories, 31 files
    ```

     **NOTE HOWEVER** that these leftover `objects` directories are empty at that point.

1. If you restart `1.10` the `SIPCreation/completedTransfers` chain reacts when it sees anything in that directory and runs the `Check transfer directory for objects` client script.

1. In `1.10` that client script and its chain link were designed to signal success when the `objects` directory of the transfer was empty and the success case pointed to a `null` terminal link. So nothing else happened, but you kept seeing the job being logged on the Transfer tab after each restart as reported in https://github.com/archivematica/Issues/issues/770

1. In `1.11` the `SIPCreation/completedTransfers` chain behaves the same (reacts by running the same client script), but the client script and its chain link have changed because of https://github.com/artefactual/archivematica/pull/1576. Now if the `objects` directory is empty the job signals `error` producing the failure report problem Miguel reports.

**Comment by sevein** [comment]
If we can confirm that we don't need the contents sitting in completedTransfers then we can suggest to remove them as part of the software upgrade. This would only be needed when upgrading from 1.10 or older, since we believe that 1.11 fixed https://github.com/archivematica/Issues/issues/769 (to be confirmed).

**Comment by tw4l** [comment]
Metadata files are no longer written to `completedTransfers` in Archivematica 1.11 as a result of commit https://github.com/artefactual/archivematica/commit/3e52494735ebfeb0cabc477d95d692034f4b3142.

In Archivematica 1.10 and before, metadata files were written into completedTransfers so that they could later be copied into the SIP by the `copy_transfer_submission_documentation.py` and `copy_transfers_metadata_and_logs.py` client scripts during the "Process submission documentation" and "Process metadata directory" microservices. The files in completedTransfers were not cleaned up after that point, but can be safely deleted after those microservices complete.

In AM 1.11, these metadata files are written into a temporary directory instead and the client scripts were modified to accordingly, so no client scripts rely on the files in `completedTransfers`, and no files are left in `completedTransfers` after a transfer is moved to the backlog or sent on to SIP creation/Ingest.

I did some testing in my local Docker dev environment and confirmed that this means we can safely delete the contents of `completedTransfers` as the first step of the upgrade procedure to 1.11 - doing so does not have an effect on transfers currently stored in the backlog, or on SIPs that have completed processing.

There is one exception: deleting the contents of the `completedTransfer` directory during the upgrade to Archivematica 1.11 **will** cause an issue is if there are SIPs that are still processing and have not yet passed the "Process submission documentation" and "Process metadata directory" microservices (e.g. a SIP that is waiting for user feedback to approve normalization or add final metadata). However, it seems reasonable to tell Archivematica users to wait until SIPs have finished processing before starting an upgrade. This is especially true because the changes made in the commit above mean that metadata files won't be successfully written into these SIPs after the upgrade to Archivematica 1.11 regardless (since the 1.11 client scripts will look for the files in `tmp`, not in `completedTransfers`).

Deleting the contents of `completedTransfers` prior to doing the upgrade to Archivematica 1.11 prevents the email failure notifications because the `SIPCreation/completedTransfers` chain never runs the `Check transfer directory for objects` client script if there's nothing in the watched directory.

I agree with @sevein that we should add a note about this to the release notes and upgrade documentation, and I will do that as my next step.

@mamedin, is deleting the contents of the `completedTransfers` watched directory something you would like to automate? Do you have thoughts on how best to do that (e.g. is this something that belongs in an Ansible playbook)?

**Comment by tw4l** [comment]
**TL;DR version:** When upgrading from Archivematica 1.10 or older to Archivematica 1.11, you can do the following to prevent these emails from being sent:

1. Check that there are no transfers or ingests that are still processing. If there are, finish processing them until the transfer or AIP is stored.
2. Delete all directories in `completedTransfers/`.
3. Perform the upgrade to AM 1.11.

**Comment by sallain** [comment]
Closing this issue as we have addressed it via documentation and the release notes. This issue is particular to the 1.10 -> 1.11 migration pathway so it is not expected to reoccur with future upgrades.
